### PR TITLE
[PM-12023] Enforce hmac when mac key is present

### PR DESCRIPTION
**Pull request by quexten**
## 🎟️ Tracking

https://bitwarden.atlassian.net/issues/PM-12023
(https://bitwarden.atlassian.net/issues/PM-4185)

## 📔 Objective

Prevent downgrading of encstring types to hmac-less by enforcing hmac when the symmetric key used has a mac key.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation
  team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: crates/bitwarden-crypto/src/enc_string/symmetric.rs [hunk 1]
```diff
@@ -237,6 +237,10 @@
     fn decrypt_with_key(&self, key: &SymmetricCryptoKey) -> Result<Vec<u8>> {
         match self {
             EncString::AesCbc256_B64 { iv, data } => {
+                if key.mac_key.is_some() {
+                    return Err(CryptoError::InvalidKey.into());
```
[the remark this answers is not in the document]
Comment (quexten) [reply crates/bitwarden-crypto/src/enc_string/symmetric.rs]: Fixed.

File: crates/bitwarden-crypto/src/enc_string/symmetric.rs [hunk 2]
```diff
@@ -418,6 +424,21 @@ mod tests {
...
+        // removed
+        let key = "hvBMMb1t79YssFZkpetYsM3deyVuQv4r88Uj9gvYe0+G8EwxvW3v1iywVmSl61iwzd17JW5C/ivzxSP2C9h7Tw==".to_string();
+        let key = SymmetricCryptoKey::try_from(key).unwrap();
+
+        let enc_str = "0.NQfjHLr6za7VQVAbrpL81w==|wfrjmyJ0bfwkQlySrhw8dA==";
+        let enc_string: EncString = enc_str.parse().unwrap();
+        assert_eq!(enc_string.enc_type(), 0);
+
+        let result: Result<String, CryptoError> = enc_string.decrypt_with_key(&key);
+        assert!(result.is_err());
```
Comment (mzieniukbw) [inline crates/bitwarden-crypto/src/enc_string/symmetric.rs:439]: could we check thart error is `CryptoError::InvalidKey` ?

File: crates/bitwarden-crypto/src/enc_string/symmetric.rs [hunk 3]
```diff
@@ -237,6 +237,10 @@ impl KeyDecryptable<SymmetricCryptoKey, Vec<u8>> for EncString {
     fn decrypt_with_key(&self, key: &SymmetricCryptoKey) -> Result<Vec<u8>> {
         match self {
             EncString::AesCbc256_B64 { iv, data } => {
+                if key.mac_key.is_some() {
+                    return Err(CryptoError::InvalidKey);
```
Comment (Hinton) [inline crates/bitwarden-crypto/src/enc_string/symmetric.rs:241]: I think we should make an explicit error about this to provide more appropriate error messages.

Maybe `KeyMissingMac`?

File: crates/bitwarden-crypto/src/enc_string/symmetric.rs [hunk 4]
```diff
@@ -418,6 +424,21 @@ mod tests {
...
+        // removed
+        let key = "hvBMMb1t79YssFZkpetYsM3deyVuQv4r88Uj9gvYe0+G8EwxvW3v1iywVmSl61iwzd17JW5C/ivzxSP2C9h7Tw==".to_string();
+        let key = SymmetricCryptoKey::try_from(key).unwrap();
+
+        let enc_str = "0.NQfjHLr6za7VQVAbrpL81w==|wfrjmyJ0bfwkQlySrhw8dA==";
+        let enc_string: EncString = enc_str.parse().unwrap();
+        assert_eq!(enc_string.enc_type(), 0);
+
+        let result: Result<String, CryptoError> = enc_string.decrypt_with_key(&key);
+        assert!(matches!(result, Err(CryptoError::InvalidKey)));
```
Comment (Hinton) [inline crates/bitwarden-crypto/src/enc_string/symmetric.rs:439]: We can improve the clarity of the comments slightly, and can we add the "real" enc string i.e. with mac here commented out?
```suggestion
        // Simulate a potential downgrade attack by removing the mac portion of the `EncString` and
        // attempting to decrypt it using a `SymmetricCryptoKey` with a mac key.
        let key = "hvBMMb1t79YssFZkpetYsM3deyVuQv4r88Uj9gvYe0+G8EwxvW3v1iywVmSl61iwzd17JW5C/ivzxSP2C9h7Tw==".to_string();
        let key = SymmetricCryptoKey::try_from(key).unwrap();

        // A "downgraded" `EncString` from `EncString::AesCbc256_HmacSha256_B64` (2) to
        // `EncString::AesCbc256_B64` (0), with the mac portion removed.
        // <enc_string>
        let enc_str = "0.NQfjHLr6za7VQVAbrpL81w==|wfrjmyJ0bfwkQlySrhw8dA==";
        let enc_string: EncString = enc_str.parse().unwrap();
        assert_eq!(enc_string.enc_type(), 0);

        let result: Result<String, CryptoError> = enc_string.decrypt_with_key(&key);
        assert!(matches!(result, Err(CryptoError::InvalidKey)));
```

**Review by Hinton — APPROVED** [review]

**Pull request merged**